docs/environment-variables.md
---
layout: default
title: Environment variables
nav_order: 2
---

# Environment variables

| Variable | Type | Required | Default | Description |
| -- | -- | -- | -- | -- |
| `NODE_ENV` | string | optional | `production` | Possible values are `development` or `production`|
| `PORT` | number | optional | `3000` |   |
| `TURBO_TOKEN` | string | optional |  | Secret token used for the authentication. Required if `AUTH_MODE` is undefined or `static`. You can specify multiple tokens separated by comma (e.g. `TURBO_TOKEN=token1,token2,token3`). The value must be the same one provided for the `token` parameter of the `build` script. See enable [custom remote caching](https://ducktors.github.io/turborepo-remote-cache/custom-remote-caching) in a Turborepo monorepo |
| `AUTH_MODE` | string | optional | `static` | Which authentication mode to use, possible values are `static` or `jwt`|
| `JWKS_URL` | string | optional | | JWKS metadata url for retrieving public keys for verifying JWTs|
| `JWT_ISSUER` | string | optional | | JWT Issuer, optional even if using JWT authentication, to match `iss` field in JWT.
| `JWT_AUDIENCE` | string | optional | | JWT Audience, optional even if using JWT authentication, to match `aud` field in JWT.
| `JWT_READ_SCOPES` | string | optional | | If specified, one of the scopes listed here must be present in order to read from the cache. You can specify multiple options with a comma-delimited string of scopes.
| `JWT_WRITE_SCOPES` | string | optional | | If specified, one of the scopes listed here must be present in order to write to the cache. You can specify multiple options with a comma-delimited string of scopes.
| `LOG_LEVEL` | string | optional | `'info'` | Possibile values are [one of these](https://github.com/ducktors/turborepo-remote-cache/blob/main/src/logger.ts#L3) |
| `ENABLE_STATUS_LOG` | boolean | optional | `'true'` | Enable/Disable logging for the status endpoint |
| `LOG_MODE` | string | optional | `stdout` | Setting it to 'file' enables writing logs to file |
| `LOG_FILE` | string | optional | `server.log` | Path and file name where save .log file (e.g. /path/to/my/file.log) |
| `STORAGE_PROVIDER` | string | optional | `local` | Possible values are `local`, `s3`, `google-cloud-storage` or `azure-blob-storage`. Use this var to choose the storage provider. |
| `STORAGE_PATH` | string | optional |  | Caching folder under `/tmp` if `STORAGE_PROVIDER` is set to `local`. If `STORAGE_PROVIDER` is set to `s3`, `google-cloud-storage` or `azure-blob-storage`, this will be the name of the bucket. |
| `STORAGE_PATH_USE_TMP_FOLDER` | boolean | optional | `true` | Uses the system tmp folder as a prefix to `STORAGE_PATH` |
| `BODY_LIMIT` | number | optional | `104857600` | The limit for artifact upload size  |
| `HTTP2` | boolean | optional | `'false'` | If set to `true`, the server will use the HTTP/2 protocol, which helps bypass the 32MB payload size limit in Cloud Run |
| `SSL_KEY_PATH` | string | optional | `` | If set, enables HTTPS using the key file at the specified path. |
| `SSL_CERT_PATH` | string | optional | `` | If set, enables HTTPS using the certificate file at the specified path. |
| `TURBO_REMOTE_CACHE_SIGNATURE_KEY` | string | optional | | A secret key used by Turborepo to sign and verify artifacts. Enables artifact integrity and authenticity verification. |

Both `SSL_KEY_PATH` and `SSL_CERT_PATH` must be set to enable HTTPS.

src/env.ts
import { Static, Type } from '@sinclair/typebox'
import Ajv from 'ajv'
import { envSchema } from 'env-schema'

const NODE_ENVS = {
  PRODUCTION: 'production',
  DEVELOPMENT: 'development',
  TEST: 'test',
} as const
type NODE_ENVS = typeof NODE_ENVS[keyof typeof NODE_ENVS]

export const STORAGE_PROVIDERS = {
  LOCAL: 'local',
  S3: 'S3',
  s3: 's3',
  GOOGLE_CLOUD_STORAGE: 'google-cloud-storage',
  AZURE_BLOB_STORAGE: 'azure-blob-storage',
  MINIO: 'minio',
} as const
export type STORAGE_PROVIDERS =
  typeof STORAGE_PROVIDERS[keyof typeof STORAGE_PROVIDERS]

const schema = Type.Object(
  {
    NODE_ENV: Type.Optional(
      Type.Enum(NODE_ENVS, { default: NODE_ENVS.PRODUCTION }),
    ),
    AUTH_MODE: Type.Optional(Type.Enum({ static: 'static', jwt: 'jwt' })),
    JWT_AUDIENCE: Type.Optional(Type.String({ separator: ',' })),
    JWT_ISSUER: Type.Optional(Type.String()),
    JWKS_URL: Type.Optional(Type.String()),
    JWT_READ_SCOPES: Type.Optional(Type.String({ separator: ',' })),
    JWT_WRITE_SCOPES: Type.Optional(Type.String({ separator: ',' })),
    TURBO_TOKEN: Type.Optional(Type.String({ separator: ',' })),
    PORT: Type.Number({ default: 3000 }),
    LOG_LEVEL: Type.Optional(Type.String({ default: 'info' })),
    ENABLE_STATUS_LOG: Type.Optional(Type.Boolean({ default: true })),
    LOG_MODE: Type.Optional(Type.String({ default: 'stdout' })),
    LOG_FILE: Type.Optional(Type.String({ default: 'server.log' })),
    STORAGE_PROVIDER: Type.Optional(
      Type.Enum(STORAGE_PROVIDERS, { default: STORAGE_PROVIDERS.LOCAL }),
    ),
    BODY_LIMIT: Type.Optional(Type.Number({ default: 104857600 })),
    STORAGE_PATH: Type.Optional(Type.String()),
    STORAGE_PATH_USE_TMP_FOLDER: Type.Optional(Type.Boolean({ default: true })),
    HTTP2: Type.Optional(Type.Boolean({ default: false })),
    // S3_ env vars are used by Vercel. ref: https://vercel.com/support/articles/how-can-i-use-aws-sdk-environment-variables-on-vercel
    S3_ACCESS_KEY: Type.Optional(Type.String()),
    S3_SECRET_KEY: Type.Optional(Type.String()),
    S3_REGION: Type.Optional(Type.String()),
    // S3_ENDPOINT is shared between are deployments type
    S3_ENDPOINT: Type.Optional(Type.String()),
    // S3_MAX_SOCKETS is used to increase the number of sockets for the S3 client
    S3_MAX_SOCKETS: Type.Optional(Type.Number({ default: 50 })),

    // Google Cloud Storage credentials
    GCS_PROJECT_ID: Type.Optional(Type.String()),
    GCS_CLIENT_EMAIL: Type.Optional(Type.String()),
    GCS_PRIVATE_KEY: Type.Optional(Type.String()),

    // Azure Blob Storage credentials
    ABS_CONNECTION_STRING: Type.Optional(Type.String()),

    // SSL support
    SSL_KEY_PATH: Type.Optional(Type.String()),
    SSL_CERT_PATH: Type.Optional(Type.String()),

    // Artifact signature
    TURBO_REMOTE_CACHE_SIGNATURE_KEY: Type.Optional(Type.String()),
  },
  {
    additionalProperties: false,
  },
)

export type Config = Static<typeof schema>
let _env: Config
export function load(overrides?: Partial<Config>) {
  _env = envSchema<Static<typeof schema>>({
    // we call the default method because Ajv provides wrong types. ref https://github.com/ajv-validator/ajv/issues/2132
    ajv: new Ajv.default({
      removeAdditional: true,
      useDefaults: true,
      coerceTypes: true,
      keywords: ['kind', 'RegExp', 'modifier', envSchema.keywords.separator],
    }),
    data: overrides,
    dotenv: process.env.NODE_ENV === NODE_ENVS.DEVELOPMENT ? true : false,
    schema,
  })
  return _env
}
_env = load()

// we export an object so we can mock the env value while testing. In fact exported vars in are not mockable in ESM
export const env = {
  get() {
    return _env
  },
}
